Kraken: fix schedules for lines not circulating on two consecutive days or more

eturck opened the pull request:
This is following issue #2279.

This problem is occurring on stop/route_schedules with a duration of multiple days. Our use case is that we want to know which lines have a service on the next seven days at a certain stop. So we ask stop_schedules the next 2 departures in the next 7 days.
Some of our lines have routes only active on Saturdays or Sundays, so when we test it at the beginning of the week, we have no results.

After @TeXitoi's answer I took the safe approach, not wanting to impact anything performance-wise. I used the already existing bound for next/previous_valid_discrete (I added it to the frequency functions).
It's not just used to stop searching after a certain bound now, I also use it to determine how far into the future I need to go. For example next_valid_discrete will search on the current day and increment the date until we pass the bound as long as we have not found the next stop time.

If the bound is not set, I set it to DateTime(dt.date + 2, 0). It's kind of "random", but it's to emulate the current behavior. I don't know if there is a more elegant way to do it. I've added a bunch of tests too.

Let me know if that's seems an OK way to fix it, if I miss something or if I need to change / add anything.

Commit:
Kraken: add unit test on next_stop_time / schedules

Add some unit test failing with long duration on stop_schedules /
route_schedules and next_stop_time. If there is a gap of more than one
day to the next service kraken will stop returning new datetime even if
there is more after. It's because right now we are only looking for the
next departure on today and tomorrow. See issue #2279.

Commit:
Kraken: expand next departure search with bound

This change the way (next/prev)_valid_(discrete/freq) look for a
departure or arrival. We are currently only looking today and the day
after / before. Now we take advantage of the existing bound to expand
the horizon if necessary. We look on the current day and loop on each
day until the bound is reached.
If no bound is passed we keep the current implied bound of tomorrow at
midnight. We make sure to not cross the end production date.

Commit:
Kraken: set bound to max_dt in get_stop_times

Allow us to use the new bound for stop and route schedules. This way we
are looking for a departure until the duration asked is reached.

buildbotctp commented:
Can one of the admins verify this patch?

kinnou02 commented:
ok to test.

xlqian commented on source/routing/next_stop_time.cpp:
```diff
@@ -133,15 +133,18 @@ next_valid_discrete(const StopEvent stop_event,
+    while(DateTimeUtils::date(bound) > date) {
+        date++;
+        for (const auto* st: st_range) {
+            assert(dataRaptor.jp_container.get_jpp(*st) == jpp_idx);
```
this is a 100% repeated code 🐱 as the L127 can you refactorize that if possbile?

pbougue commented on source/routing/next_stop_time.cpp:
```diff
@@ -153,98 +156,71 @@ static std::pair<const type::StopTime*, DateTime>
         const dataRAPTOR& dataRaptor,
         const JppIdx jpp_idx,
-        const DateTime dt,
+        DateTime dt,
```
It will probably imply code a bit away of your PR but I'd rather add `const`s than remove'em.
But it's int underlying here, so copied...

eturck replied:
I can keep the `const` and start by copying dt in another `DateTime` if it's better. That's not really a problem I just took the easy way without much thinking ;).

pbougue replied:
Or even... propagate it 😝

eturck replied:
I'm not really sure to fully understand sorry :worried:. Do want me to just keep the `const` and work on a copy of the `DateTime` and also add a `const` to the dt of `get_next_stop_time`? Or something else/more?

pbougue replied:
Probably work on a copy for the needed part (as it is mutable, which I didn't remark at first sight).
And also at first my idea was to add `const`s to function params you are using (`DateTimeUtils::date()` and `get_next_stop_time()`): it's legacy, you can do it if you want but it's not your responsibility at all :)

pbougue reviewed (approved):
Looks good
I'm just wondering if we might hit some "DOS" issues with people going a bit too far in time with big counts, or things like that.

eturck commented:
@pbougue it could probably cause some issue with people getting a stop_schedules for a year or something. Maybe like there is a max_count in the config you could have a max_duration too? It's not perfect if everybody can request something like two weeks but it's limited compare to the end production date as a limit.

Commit:
Kraken: refactor next/prev_valid_discrete to reduce repeated code

Also add consts back on frequencies and added them on inline methods of
DateTimeUtils.

eturck commented:
I changed next/previous_valid_discrete to reduce repeated code and added some const :).
